pysatMissions/methods/spacecraft.py

```diff
@@ -0,0 +1,235 @@
+
+    inst.meta['sc_yhat_ecef_x'] = {'units': '',
+                                   'desc': 'S/C attitude (y-direction, ' +
+                                   'generally south) unit vector, expressed ' +
+                                   'in ECEF basis, x-component'}
+    inst.meta['sc_yhat_ecef_y'] = {'units': '',
+                                   'desc': 'S/C attitude (y-direction, ' +
+                                   'generally south) unit vector, expressed ' +
+                                   'in ECEF basis, y-component'}
+    inst.meta['sc_yhat_ecef_z'] = {'units': '',
+                                   'desc': 'S/C attitude (y-direction, ' +
+                                   'generally south) unit vector, expressed ' +
+                                   'in ECEF basis, z-component'}
+
+    # check what magnitudes we get
+    mag = np.sqrt(inst['sc_zhat_ecef_x']**2 + inst['sc_zhat_ecef_y']**2 +
+                  inst['sc_zhat_ecef_z']**2)
+    idx, = np.where((mag < .999999999) | (mag > 1.000000001))
+    if len(idx) > 0:
+        print(mag[idx])
+        raise RuntimeError('Unit vector generation failure. Not sufficently ' +
+                           'orthogonal.')
+
+    return
+
+
+def calculate_ecef_velocity(inst):
+    """
+    Calculates spacecraft velocity in ECEF frame.
+
+    Presumes that the spacecraft velocity in ECEF is in
+    the input instrument object as position_ecef_*. Uses a symmetric
+    difference to calculate the velocity thus endpoints will be
+    set to NaN. Routine should be run using pysat data padding feature
+    to create valid end points.
+
+    Parameters
+    ----------
+    inst : pysat.Instrument
+        Instrument object
+
+    Returns
+    -------
+    None
+        Modifies pysat.Instrument object in place to include ECEF velocity
+        using naming scheme velocity_ecef_* (*=x,y,z)
+
+    """
+
+    def get_vel_from_pos(x):
+        vel = (x.values[2:] - x.values[0:-2])/2.
+        return vel
+
+    vel_x = get_vel_from_pos(inst['position_ecef_x'])
+    vel_y = get_vel_from_pos(inst['position_ecef_y'])
+    vel_z = get_vel_from_pos(inst['position_ecef_z'])
+
+    inst[1:-1, 'velocity_ecef_x'] = vel_x
+    inst[1:-1, 'velocity_ecef_y'] = vel_y
+    inst[1:-1, 'velocity_ecef_z'] = vel_z
+
+    inst.meta['velocity_ecef_x'] = {'units': 'km/s',
+                                    'desc': 'Velocity of satellite ' +
+                                    'calculated with respect to ECEF frame.'}
+    inst.meta['velocity_ecef_y'] = {'units': 'km/s',
+                                    'desc': 'Velocity of satellite ' +
+                                    'calculated with respect to ECEF frame.'}
+    inst.meta['velocity_ecef_z'] = {'units': 'km/s',
+                                    'desc': 'Velocity of satellite ' +
+                                    'calculated with respect to ECEF frame.'}
+    return
+
+
+def project_ecef_vector_onto_sc(inst, x_label, y_label, z_label,
```

Comment:
This method could use some generalization. Outside scope of this pool, will create an issue.

Reply:
Thanks for writing this up!